### PR TITLE
Docker v2plugin support

### PR DESCRIPTION
**Pull request by gkvijay**
+ Makefile targets (demo-v2plugin) for the complete workflow
  and for subtasks like creating/restarting plugin
+ Stubs for new libnetwork apis
+ Changes to use the name of v2plugin in docknet apis
+ Scripts to create the v2plugin
+ Python supporting scripts for v2plugin start/stop
+ Updated docker and libnetwork libraries

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -67,7 +68,18 @@ else
...
+    if [[ #{docker_version} == *"rc"* ]]; then
+        echo "Getting pre-release docker version #{docker_version} "
+        curl -fsSL https://test.docker.com/ | sh
+    elif [[ #{docker_version} == *"cs"* ]]; then
+            echo "Getting cs docker version #{docker_version} "
+            curl -fsSL https://packages.docker.com/1.13/install.sh | sh
+    else
+        echo "Getting released docker version #{docker_version} "
+        curl https://get.docker.com | sed s/docker-engine/docker-engine-#{docker_version}/ | bash
+            echo "Getting released docker version #{docker_version} "
```
Comment (jojimt) [inline Vagrantfile]: These lines repeated by mistake?

File: Vagrantfile [hunk 2]
```diff
@@ -298,6 +310,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
                   ENV["CONTIV_CLUSTER_STORE"] || "etcd://localhost:2379",
                   ENV["CONTIV_DOCKER_VERSION"] || docker_version,
                   ENV['CONTIV_NODE_OS'] || "",
+                  ENV['CONTIV_V2PLUGIN_NAME'] || "contiv/v2netplugin:0.1",
```
Comment (jojimt) [inline Vagrantfile]: I think we should probably use the netplugin version string as tag, instead of 0.1. But, this is ok for now.

File: install/v2plugin/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+FROM ubuntu:16.04
+MAINTAINER Cisco Contiv (http://contiv.github.io/)
+
+RUN mkdir -p /run/docker/plugins
+RUN mkdir -p /etc/openvswitch
+RUN mkdir -p /var/run/contiv/log
+
+RUN apt-get update
+RUN apt-get --assume-yes  install openvswitch-switch
```
Comment (jojimt) [inline install/v2plugin/Dockerfile]: We need to specify the ovs version here. Please check the latest install scripts. @neelimamukiri had added this.

File: mgmtfn/dockplugin/netDriver.go [hunk 4]
```diff
@@ -371,6 +371,58 @@ func leave(w http.ResponseWriter, r *http.Request) {
...
+func networkAllocate(w http.ResponseWriter, r *http.Request) {
+	resp := api.AllocateNetworkResponse{}
+
+	content, err := json.Marshal(resp)
+	if err != nil {
+		httpError(w, "Could not generate a response", err)
+		return
+	}
+
+	w.Write(content)
```
Comment (jojimt) [inline mgmtfn/dockplugin/netDriver.go]: Is this just a stub function? Can we eliminate the redundant return statement?

File: mgmtfn/dockplugin/netDriver.go [hunk 5]
```diff
@@ -371,6 +371,58 @@ func leave(w http.ResponseWriter, r *http.Request) {
...
+	return
+}
+
```
Comment (jojimt) [inline mgmtfn/dockplugin/netDriver.go]: Same comment for all the above routines.

File: netmaster/main.go [hunk 6]
```diff
@@ -23,12 +23,14 @@ import (
...
 	log "github.com/Sirupsen/logrus"
 	"github.com/contiv/netplugin/netmaster/daemon"
+	"github.com/contiv/netplugin/netmaster/docknet"
 	"github.com/contiv/netplugin/version"
 )
 
 type cliOpts struct {
 	help         bool
 	debug        bool
+	pluginName   string
```
Comment (jojimt) [inline netmaster/main.go]: Is there a need to pass the plugin name?

File: Godeps/Godeps.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
 	"ImportPath": "github.com/contiv/netplugin",
 	"GoVersion": "go1.7",
-	"GodepVersion": "v78",
+	"GodepVersion": "v75",
```
Comment (jojimt) [inline Godeps/Godeps.json]: You should probably update your godep version.

File: Vagrantfile (hunk 1 above)
Comment (gkvijay) [inline Vagrantfile]: location to fetch the rc (release candidate) and cs (commercial support) versions of docker are different from the released versions

File: Vagrantfile (hunk 2 above)
Comment (gkvijay) [inline Vagrantfile]: Just put the default name as contiv/netplugin:0.1, but we can override this with the env variable. Will update the release scripts to create/release plugin with release version tag as a new PR

File: mgmtfn/dockplugin/netDriver.go (hunk 4 above)
Comment (gkvijay) [inline mgmtfn/dockplugin/netDriver.go]: Eventhough we dont do any processing of these APIs, we need to return a valid response to avoid failure in dockerd. Will remove the return

File: mgmtfn/dockplugin/netDriver.go (hunk 5 above)
Comment (gkvijay) [inline mgmtfn/dockplugin/netDriver.go]: will remove the redundant return

File: netmaster/main.go (hunk 6 above)
Comment (gkvijay) [inline netmaster/main.go]: netmaster needs to know the plugin name so that it can use it in the docknet apis to create the docker network using the proper plugin name

File: Godeps/Godeps.json (hunk 7 above)
Comment (gkvijay) [inline Godeps/Godeps.json]: will update godep version and push

File: install/v2plugin/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+RUN apt-get update && \
+    apt-get --assume-yes install openvswitch-switch=2.5.0-0ubuntu1 && \
+    apt-get --assume-yes install net-tools iptables && \
+    apt-get --assume-yes install iputils-ping netcat
+
+COPY netplugin netmaster netctl startcontiv.sh /
+
+ENTRYPOINT ["/startcontiv.sh"]
+#CMD ["/startcontiv.sh"]
```
Comment (rhim) [inline install/v2plugin/Dockerfile]: Remove.

File: Vagrantfile (hunk 2 above)
Comment (rhim) [inline Vagrantfile]: Why use a separate versioning scheme for v2netplugin container image? Shouldn't we use the same version as what is in netplugin/version/CURRENT_VERSION?

File: install/v2plugin/Dockerfile (hunk 8 above)
Comment (gkvijay) [inline install/v2plugin/Dockerfile]: will do

File: Vagrantfile (hunk 2 above)
Comment (gkvijay) [inline Vagrantfile]: this is a default version for dev and the plugin is not pushed to hub. In the release script we should use the correct version and push it to the hub. Will add that in a following PR. Need to discuss that with you, joji and neelima

Comment (gkvijay) [inline Vagrantfile]: Need to decide the official name of released plugin as well

File: Vagrantfile (hunk 1 above)
Comment (jojimt) [inline Vagrantfile]: True, but you have the echo and curl repeated twice in this else block...

Comment (gkvijay) [inline Vagrantfile]: Thanks. I missed that. Removed the repeated lines

File: install/v2plugin/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# Docker v2plugin container with OVS / netplugin / netmaster 
+
+FROM ubuntu:16.04
+MAINTAINER Cisco Contiv (http://contiv.github.io/)
+
+RUN mkdir -p /run/docker/plugins /etc/openvswitch /var/run/contiv/log
```
Comment (unclejack) [inline install/v2plugin/Dockerfile]: Please change this and the lines below to the following:
```
RUN mkdir -p /run/docker/plugins /etc/openvswitch /var/run/contiv/log \
 && apt-get update \
 && apt-get install -y openvswitch-switch=2.5.0-0ubuntu1 \
        net-tools iptables iputils-ping netcat \
 && rm -f /var/lib/apt/lists/*
```

Comment (gkvijay) [inline install/v2plugin/Dockerfile]: Updated with the change

File: Vagrantfile [hunk 10]
```diff
@@ -67,7 +68,16 @@ else
     # cleanup openstack-kilo repo if required
     yum remove docker-engine -y || :
     yum-config-manager --disable openstack-kilo
-    curl https://get.docker.com | sed s/docker-engine/docker-engine-#{docker_version}/ | bash
+    if [[ #{docker_version} == *"rc"* ]]; then
```
Comment (unclejack) [inline Vagrantfile]: These changes will become redundant soon. The Docker engine versioning has changed. The `-cs` edition won't exist any more. The regular Docker engine version will become `-ce` (community edition).

Can we skip these changes for now, please?

File: Godeps/Godeps.json [hunk 11]
```diff
@@ -128,13 +128,18 @@
 		},
 		{
 			"ImportPath": "github.com/docker/docker/pkg/plugins",
-			"Comment": "v1.4.1-13296-ge904f58",
-			"Rev": "e904f58e199836fa43ee1c6736e689ef58e4fe10"
+			"Comment": "docs-v1.12.0-rc4-2016-07-15-3221-gccba791",
```
Comment (unclejack) [inline Godeps/Godeps.json]: Why is the rc4 version of Go used here? We should use stable versions.

File: Godeps/Godeps.json [hunk 12]
```diff
@@ -128,13 +128,18 @@
 		},
```
Comment (unclejack) [inline Godeps/Godeps.json]: All the Godeps changes should be in their own commit. This will help us track down potential issues caused by this change.

The commit which updates these libraries should be the first one. This PR is quite big and we should split it into multiple commits.

File: Vagrantfile (hunk 10 above)
Comment (gkvijay) [inline Vagrantfile]: Since we are working with pre-released docker versions as well, we need the rc version. cs is the customer support version which was suggested by docker folks.

File: Godeps/Godeps.json (hunk 12 above)
Comment (jojimt) [inline Godeps/Godeps.json]: @unclejack I don't think what you're asking is always possible. Many times godep update will require corresponding changes to just make things compile.

Comment (dseevr) [inline Godeps/Godeps.json]: @jojimt I think he means:

```sh
make godep-save
git add Godeps
git add vendor
git commit -m "updated Godeps"
```

Just so that it generates a separate commit, not a separate PR.  As an example, I just did this here:  https://github.com/contiv/objdb/pull/20

Comment (gkvijay) [inline Godeps/Godeps.json]: Split it into two commits as suggested by @unclejack

File: install/v2plugin/Dockerfile (hunk 3 above)
Comment (gkvijay) [inline install/v2plugin/Dockerfile]: updated the ovs version in Dockerfile

File: Vagrantfile (hunk 10 above)
Comment (unclejack) [inline Vagrantfile]: The `-rc` thing should be OK. The `-cs` part will go away soon. There's no point in adding something for the `-cs` versions of Docker at this point. https://github.com/docker/docker/pull/31075

File: Godeps/Godeps.json [hunk 13]
```diff
@@ -624,6 +629,111 @@
 			"ImportPath": "github.com/coreos/go-semver/semver",
 			"Comment": "v0.2.0-5-g5e3acbb",
 			"Rev": "5e3acbb5668c4c3deb4842615c4098eb61fb6b1e"
+		},
+		{
+			"ImportPath": "github.com/docker/docker/vendor/github.com/Sirupsen/logrus",
```
Comment (rchirakk) [inline Godeps/Godeps.json]: Is the nested vendoring going to create multiple versions of logrus/other packages in netplugin?

File: netmaster/main.go [hunk 14]
```diff
@@ -90,6 +96,8 @@ func execOpts(opts *cliOpts) {
 	if opts.debug {
 		log.SetLevel(log.DebugLevel)
 	}
+
+	docknet.UpdatePluginName(opts.pluginName)
```
Comment (rchirakk) [inline netmaster/main.go]: do you need to check if netmaster is started in docker mode here ?

Comment (gkvijay) [inline netmaster/main.go]: Here the plugin name is being updated by netmaster when it starts

File: Godeps/Godeps.json (hunk 13 above)
Comment (gkvijay) [inline Godeps/Godeps.json]: updated to avoid nested vendoring by picking specific packages

**Comment by unclejack** [comment]
The tests pass for the regular netplugin setup without the v2 plugin.

I'll try to run the systemtests with the v2 plugin.

**Comment by unclejack** [comment]
@gkvijay: This doesn't hook into test/systemtests. This is required to run the tests with this plugin.

**Comment by gkvijay** [comment]
@unclejack, v2plugin is not enabled for systemtest. That will be tested/fixed as a separate PR. We need an alpha quality plugin build ready for plugin certification with docker
